### DISCUSSION TITLE
Bootstrap improvement

### DISCUSSION THREAD

**Discussion by Valgrifer** [PaperMC/Paper#11289] [Ideas]
### Is your feature request related to a problem?

Lack of function in the context of bootstrap and event. in the case of generating a resource pack linked with a datapack

### Describe the solution you'd like.

An event or function that would be triggered after all the "bootstrap" functions of all plugins and before the server launch.
Access to the ClassLoader to have the resources from the bootstrap.
And access to the config simply

### Describe alternatives you've considered.

I find that these are the minimum features to manage resources before launching the server and which will be useful in the future when we can add all types of data-driven content from the bootstrap.

**Comment by Machine-Maker** [comment]
No doubt there is functionality missing. 

Can you expand on why you want these specific things? Like what is the [xy](https://xyproblem.info/) here?

> An event or function that would be triggered after all the "bootstrap" functions of all plugins and before the server launch.

What happens during the bootstrap that you have to wait until *after*?

> Access to the ClassLoader to have the resources from the bootstrap.

which classloader? You can easily get the classloader for any class with `WhateverClass.class.getClassLoader()`

> And access to the config simply

The config found in the plugin jar? You can also read that yourself using available methods to get resources, just like one would with any *custom* configuration file in normal plugin operations. Or are you talking about some other config?

**Reply by Valgrifer** [reply]
> What happens during the bootstrap that you have to wait until after?

In my case, gather all the resources requested by other plugins to generate (if necessary) a resource pack to distribute to players afterwards

> which classloader? You can easily get the classloader for any class with WhateverClass.class.getClassLoader()

The ClassLoader that we find in JavaPlugin with `getClassLoader()`, I am so used to using this method that I had forgotten this way of doing it

> The config found in the plugin jar? You can also read that yourself using available methods to get resources, just like one would with any custom configuration file in normal plugin operations. Or are you talking about some other config?

The plugin config in `plugins/MyPlugin/config.yml` and the server config found in `Server.Spigot`

**Reply by masmc05** [reply]
> In my case, gather all the resources requested by other plugins to generate (if necessary) a resource pack to distribute to players afterwards

Why can't you do this in onEnable? The players still won't be on the server until then so you can generate the resource pack and use it when the players will join

> The plugin config in plugins/MyPlugin/config.yml and the server config found in Server.Spigot

For your config just '[YamlConfiguration.loadConfiguration](<https://jd.papermc.io/paper/1.21.1/org/bukkit/configuration/file/YamlConfiguration.html#loadConfiguration(java.io.File)>)(context.[getDataDirectory()](<https://jd.papermc.io/paper/1.21.1/io/papermc/paper/plugin/bootstrap/PluginProviderContext.html#getDataDirectory()>).resolve("config.yml").toFile())', server config may depend in future on registry values, which are available only post bootstrap, so I don't think it's a good idea to expose them.
